### PR TITLE
Add bandit and pylama with the additional linters mypy and eradicate.

### PR DESCRIPTION
**Pull request by real-yfprojects**
Follow up on https://github.com/johannesjh/req2flatpak/pull/26#issuecomment-1296303126.

* pyproject.toml (tool.poetry.group.lint): Add dependencies `pylama[mypy,pylint,eradicate]` and `bandit`.
	Remove `pylint` dep.

* .pre-commit-config.yaml : Configure bandit and pylama. Remove pylint hook.

* pyproject.toml : Configure the added linters.

* setup.cfg : Configure stand-alone `pycodestyle`.

* poetry.lock : Update.

### CODE REVIEW

**Comment by real-yfprojects** [comment]
Only one issue left: Loading untrusted data with the `shelve` module is a security flaw since it allows implanting custom code.

**Comment by johannesjh** [comment]
About `shelve`: I am aware of that, but I do not consider it a vulnerability or threat:
* about the shelf's payload: the shelf is created by our own req2flatpak script. and yes, the payload is external content from pypi. but it is a serialized json string, and I think there is good reason to believe that this serialization is safe.
* about the the shelf file: yes, the shelf file could be maliciously manipulated on the local computer. but this would mean the computer is already compromised. it is in principle impossible to run code securely on a compromised platform. and from a more practical perspective: if someone or something can already manipulate the local shelf file, then he/she/it already has better attack possibilities than to rely on req2flatpak's caching mechanism.

It is therefore not strictly necessary in my opinion to replace the shelf with another caching mechanism. (if you still want to do it or feel that it should be done, I don't object).

File: setup.cfg [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[pycodestyle]
```
Comment (johannesjh) [inline setup.cfg]: About `setup.cfg`: is this file really necessary? You also mentioned that you don't like the redundant configuration, and I totally agree, I also don't like the redundancy. I think it would be better to remove setup.cfg. 

Note that running pycodestyle in a stand-alone manner is still possibly using `pylama -l pycodestyle`

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -22,20 +22,28 @@ repos:
...
+      - id: bandit
+        files: "req2flatpak.py"
+        args:
+          - "--aggregate"
+          - "vuln"
+          - "-ll"
+          - "-q"
+          - "-f"
+          - "custom"
+          - "--msg-template"
```
Comment (johannesjh) [inline .pre-commit-config.yaml:45]: out of interest... why a custom message template? (my gut feeling is that it may be better to rely on default message templates to avoid having to maintain and debug the template in the future)

File: req2flatpak.py [hunk 3]
```diff
@@ -457,7 +458,8 @@ def _query_from_cache(url) -> Optional[str]:
                 return None
 
         def _query_from_pypi(url) -> str:
-            with urllib.request.urlopen(url) as response:
+            # this an internal method that will be called with a trusted URL
+            with urllib.request.urlopen(url) as response:  # nosec
```
Comment (johannesjh) [inline req2flatpak.py]: a more specific ignore rule would be better (to avoid masking of other security problems that were maybe not considered). in this case: `#nosec B310` should do the job.

Comment (johannesjh) [reply req2flatpak.py]: or even better: instead of just silencing the warning, we could add a guard to ensure the url starts with `https`.

**Review by johannesjh — CHANGES_REQUESTED** [review]

**Comment by johannesjh** [comment]
hi, thanks for all the good work! I reviewed and suggested some changes using githubs "review" functionality, see above. thank you! br, Johannes

**Comment by johannesjh** [comment]
Oh, one more point: could you update the makefile with the new commands used for linting? (motivation behind the makefile: I primarily use it to avoid having to remember the commands for building / testing / running the project).

**Comment by real-yfprojects** [comment]
> he/she/it already has better attack possibilities than to rely on req2flatpak's caching mechanism

The attacker could override installed python modules directly.

File: setup.cfg (hunk 1 above)
Comment (real-yfprojects) [reply setup.cfg]: > pylama -l pycodestyle

I didn't know that.

File: req2flatpak.py (hunk 3 above)
Comment (real-yfprojects) [reply req2flatpak.py:462]: > a more specific ignore rule would be better (to avoid masking of other security problems that were maybe not considered). in this case: #nosec B310 should do the job.

Nope. Bandit only understands `# nosec`.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (real-yfprojects) [reply .pre-commit-config.yaml:45]: The default output includes run metrics and long, elaborate issue descriptions. I think that this is impractical for use as a pre-commit hook. The default template for the `custom` format option is too hard to read.

File: req2flatpak.py (hunk 3 above)
Comment (real-yfprojects) [reply req2flatpak.py]: > in this case: #nosec B310 should do the job.

That works indeed.

**Comment by real-yfprojects** [comment]
> could you update the makefile with the new commands used for linting?

There is nothing to update since the `make lint` just calls pre-commit.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (johannesjh) [reply .pre-commit-config.yaml:45]: ok alright, thanks for explaining!

**Review by johannesjh — APPROVED** [review]
looks good to me, thank you! feel free to merge.
